sector-builder/src/scheduler.rs

```diff
@@ -141,26 +133,62 @@ impl Scheduler {
-                        Ok(protos) => {
-                            for p in protos {
-                                worker_tx
-                                    .send(WorkerTask::from_seal_proto(p, scheduler_tx.clone()))
-                                    .expects(FATAL_NOSEND);
-                            }
+                    SchedulerTask::SealAllStagedSectors(tx) => {
+                        m.mark_all_sectors_for_sealing();
 
-                            tx.send(Ok(())).expects(FATAL_NOSEND);
-                        }
-                        Err(err) => {
-                            tx.send(Err(err)).expects(FATAL_NOSEND);
+                        match m.create_seal_task_protos(|x| {
+                            x.seal_status == SealStatus::ReadyForSealing
+                        }) {
+                            Ok(protos) => {
+                                for p in protos {
+                                    m.commit_sector_to_ticket(&p.sector_id, &p.seal_ticket);
+
+                                    worker_tx
+                                        .send(WorkerTask::from_seal_proto(
+                                            p.clone(),
+                                            scheduler_tx.clone(),
+                                        ))
+                                        .expects(FATAL_NOSEND);
+                                }
+
+                                tx.send(Ok(())).expects(FATAL_NOSEND);
+                            }
+                            Err(err) => {
+                                tx.send(Err(err)).expects(FATAL_NOSEND);
+                            }
                         }
-                    },
+                    }
                     SchedulerTask::SetCurrentSealTicket(seal_ticket, tx) => {
                         m.set_current_seal_ticket(seal_ticket);
-                        tx.send(Ok(())).expects(FATAL_NOSEND);
+
+                        match m.create_seal_task_protos(|x| {
+                            x.seal_status == SealStatus::ReadyForSealing
+                        }) {
+                            Ok(protos) => {
+                                for p in protos {
+                                    m.commit_sector_to_ticket(&p.sector_id, &p.seal_ticket);
+
+                                    worker_tx
+                                        .send(WorkerTask::from_seal_proto(
+                                            p.clone(),
+                                            scheduler_tx.clone(),
+                                        ))
+                                        .expects(FATAL_NOSEND);
+                                }
+
+                                tx.send(Ok(())).expects(FATAL_NOSEND);
+                            }
+                            Err(err) => {
+                                tx.send(Err(err)).expects(FATAL_NOSEND);
+                            }
+                        }
```

Comment:
Maybe extract the duplicated functionality above into a function?

Reply:
done